teps/0135-coscheduling-pipelinerun-pods.md

```diff
@@ -0,0 +1,412 @@
+
+## Design Details
+
+### Configuration
+
+Currently, the affinity assistant is enabled by default, and disabled by setting the boolean feature flag "disable-affinity-assistant" to true.
+
+This TEP proposes creating a new feature flag, "coscheduling". The options will be:
+- "disabled": The default value.
+- "coschedule-workspaces": The existing affinity assistant behavior. This option runs all of a PipelineRun's pods sharing the same workspace on the same node.
+- "coschedule-pipelineruns": This option runs all of a PipelineRun's pods on the same node.
+- "isolate-pipelineruns": This option runs all of a PipelineRun's pods on the same node, and only allows one PipelineRun to run on a node at a time.
+
+If the existing affinity assistant is enabled, the "coscheduling" flag must be set to "disabled". We will deprecate the "disable-affinity-assistant" flag in favor of the "coscheduling" flag, as discussed in ["Upgrade and Migration Strategy"](#upgrade-and-migration-strategy).
+
+### Placeholder pod + inter-pod affinity
+
+The proposed implementation is very similar to the existing affinity assistant. Using a [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/), the PipelineRun controller will create a placeholder pod per PipelineRun. All TaskRun pods will be modified to have inter-pod affinity for the placeholder pod. If the "isolate-pipelineruns" configuration option is chosen, the placeholder pod will have inter-pod anti-affinity for other placeholder pods. The placeholder pod does nothing, has minimal resource requirements, and is torn down on PipelineRun completion.
+
+#### Example: Coscheduling a PipelineRun
+
+The placeholder pod will have the following labels:
+
+```
+app.kubernetes.io/component: per-pipelinerun-affinity-assistant
+app.kubernetes.io/instance: per-pipelinerun-affinity-assistant-1ee5adc3e5
+```
+
+The following affinity terms will be added to TaskRun pods:
+```yaml
+affinity:
+  podAffinity:
+    requiredDuringSchedulingIgnoredDuringExecution:
+    - labelSelector:
+        matchLabels:
+          app.kubernetes.io/component: affinity-assistant
+          app.kubernetes.io/instance: affinity-assistant-1ee5adc3e5
+      topologyKey: kubernetes.io/hostname
+```
+
+This requires Kubernetes to schedule TaskRun pods to the same node as the placeholder pod.
+
+#### Example: Isolating PipelineRuns
+
+The only implementation difference with this configuration is that the following affinity term will be added to the placeholder pod:
+
+```yaml
+affinity:
+  podAntiAffinity:
+    requiredDuringSchedulingIgnoredDuringExecution:
+    - podAffinityTerm:
+        labelSelector:
+          matchLabels:
+            app.kubernetes.io/component: affinity-assistant
+        topologyKey: kubernetes.io/hostname
+```
+This prevents two placeholder pods from being scheduled to the same node, which in turn prevents multiple PipelineRuns from running on the same node.
```

Comment:
This configuration option would probably be chosen in an environment with a cluster autoscaler and many nodes-- so instead of PipelineRuns timing out waiting for a low number of nodes, we'd have new nodes scaled up and down for each PipelineRun. In the environment you described the cluster operator would probably choose the "coschedule-pipelineruns" option.

Reply:
+1 one of the assumptions for the isolation mode is that you are running in a cluster that can autoscale nodes